File: src/main/java/net/minestom/server/attribute/AttributeModifier.java
package net.minestom.server.attribute;

import org.jetbrains.annotations.NotNull;

import java.util.UUID;

/**
 * Represent an attribute modifier.
 */
public class AttributeModifier {

    private final float amount;
    private final String name;
    private final AttributeOperation operation;
    private final UUID id;

    /**
     * Creates a new modifier with a random id.
     *
     * @param name      the name of this modifier
     * @param amount    the value of this modifier
     * @param operation the operation to apply this modifier with
     */
    public AttributeModifier(@NotNull String name, float amount, @NotNull AttributeOperation operation) {
        this(UUID.randomUUID(), name, amount, operation);
    }

    /**
     * Creates a new modifier.
     *
     * @param id        the id of this modifier
     * @param name      the name of this modifier
     * @param amount    the value of this modifier
     * @param operation the operation to apply this modifier with
     */
    public AttributeModifier(@NotNull UUID id, @NotNull String name, float amount, @NotNull AttributeOperation operation) {
        this.id = id;
        this.name = name;
        this.amount = amount;
        this.operation = operation;
    }

    /**
     * Gets the id of this modifier.
     *
     * @return the id of this modifier
     */
    @NotNull
    public UUID getId() {
        return id;
    }

    /**
     * Gets the name of this modifier.
     *
     * @return the name of this modifier
     */
    @NotNull
    public String getName() {
        return name;
    }

    /**
     * Gets the value of this modifier.
     *
     * @return the value of this modifier
     */
    public float getAmount() {
        return amount;
    }

    /**
     * Gets the operation of this modifier.
     *
     * @return the operation of this modifier
     */
    @NotNull
    public AttributeOperation getOperation() {
        return operation;
    }
}

File: src/main/java/net/minestom/server/attribute/AttributeModifier.kt
package net.minestom.server.attribute

import java.lang.IllegalArgumentException
import java.util.concurrent.ConcurrentHashMap
import net.minestom.server.attribute.AttributeInstance
import net.minestom.server.attribute.AttributeModifier
import net.minestom.server.attribute.AttributeOperation
import java.util.*
import java.util.function.IntFunction

/**
 * Represent an attribute modifier.
 */
class AttributeModifier
/**
 * Creates a new modifier.
 *
 * @param id        the id of this modifier
 * @param name      the name of this modifier
 * @param amount    the value of this modifier
 * @param operation the operation to apply this modifier with
 */(
    /**
     * Gets the id of this modifier.
     *
     * @return the id of this modifier
     */
    val id: UUID,
    /**
     * Gets the name of this modifier.
     *
     * @return the name of this modifier
     */
    val name: String,
    /**
     * Gets the value of this modifier.
     *
     * @return the value of this modifier
     */
    val amount: Float,
    /**
     * Gets the operation of this modifier.
     *
     * @return the operation of this modifier
     */
    val operation: AttributeOperation
) {

    /**
     * Creates a new modifier with a random id.
     *
     * @param name      the name of this modifier
     * @param amount    the value of this modifier
     * @param operation the operation to apply this modifier with
     */
    constructor(name: String, amount: Float, operation: AttributeOperation) : this(
        UUID.randomUUID(),
        name,
        amount,
        operation
    ) {
    }
}